### PR TITLE
feat(ui): add multi context support to chat

### PR DESCRIPTION
**Pull request by liangfung**
https://jam.dev/c/72654039-8ede-46a5-bb82-3cac5aadaa75

### CODE REVIEW

File: ee/tabby-ui/components/chat/use-tabby-answer.ts [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+
+      if (!response?.ok) {
+        throw new Error(String(response.status))
+      }
+
+      if (response.body == null) {
+        throw new Error('The response body is empty')
+      }
+
+      for await (const line of readDataStream(response.body.getReader(), {
```
Comment (wsxiaoys) [inline ee/tabby-ui/components/chat/use-tabby-answer.ts:75]: consider using a event stream parser, @icycodes happen to investigate this area recently

File: ee/tabby-ui/components/chat/use-tabby-answer.ts [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+
+  const mergeParsedData = (
+    data1: Partial<AnswerResponse>,
+    data2: Partial<AnswerResponse>
+  ): Partial<AnswerResponse> => {
+    return mergeWith({}, data1, data2, (obj, src) => {
+      if (typeof obj === 'string' && typeof src === 'string') {
+        return obj + src
+      }
+    })
```
Comment (wsxiaoys) [inline ee/tabby-ui/components/chat/use-tabby-answer.ts:42]: the only field requires merging is `answerDelta`, consider simplify logic here.

File: ee/tabby-ui/components/chat/question-answer.tsx [hunk 3]
```diff
@@ -133,9 +144,21 @@ function AssistantMessageCard(props: AssistantMessageCardProps) {
...
-    return compact([selectContext, ...(relevantContext ?? [])])
-  }, [selectContext, relevantContext])
+  const contexts: Array<Context> = React.useMemo(() => {
+    return (
+      message?.relevant_code?.map(code => {
+        return {
+          kind: 'file',
+          range: {
+            start: code.start_line ?? 0,
+            end: code.start_line ?? 0
```
Comment (wsxiaoys) [inline ee/tabby-ui/components/chat/question-answer.tsx:154]: need compute end properly

Comment (liangfung) [reply ee/tabby-ui/components/chat/question-answer.tsx]: done

File: ee/tabby-ui/components/chat/use-tabby-answer.ts (hunk 1 above)
Comment (liangfung) [reply ee/tabby-ui/components/chat/use-tabby-answer.ts]: using 'eventsource-parser'

File: clients/tabby-chat-panel/src/index.ts [hunk 4]
```diff
@@ -10,6 +10,7 @@ export interface FileContext {
   range: LineRange
   filepath: string
   content?: string
```
Comment (wsxiaoys) [inline clients/tabby-chat-panel/src/index.ts:12]: why this is optional?

File: clients/tabby-chat-panel/src/index.ts [hunk 5]
```diff
@@ -10,6 +10,7 @@ export interface FileContext {
...
+  git_url?: string
```
Comment (wsxiaoys) [inline clients/tabby-chat-panel/src/index.ts:13]:
```suggestion
  git_url: string
```

File: ee/tabby-ui/lib/types/chat.ts [hunk 6]
```diff
@@ -40,9 +42,37 @@ export type ISearchHit = {
...
   num_hits?: number
 }
 
 export type MessageActionType = 'delete' | 'regenerate'
+
+export type CodeSearchDocument = {
+  body: string
+  filepath: string
+  git_url: string
+  language?: string
```
Comment (wsxiaoys) [inline ee/tabby-ui/lib/types/chat.ts:57]:
```suggestion
  language: string
```

File: ee/tabby-ui/lib/types/chat.ts [hunk 7]
```diff
@@ -40,9 +42,37 @@ export type ISearchHit = {
...
+  start_line?: number
```
Comment (wsxiaoys) [inline ee/tabby-ui/lib/types/chat.ts:58]:
```suggestion
  start_line: number
```

**Pull request merged**